In [1]:
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# 讀入深度學習套件

In [2]:
from tensorflow.keras.preprocessing import sequence
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Embedding
from tensorflow.keras.layers import LSTM
from tensorflow.keras.datasets import imdb

# 讀入數據

Number of words=12,000

In [3]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=12000)

# 資料處理

資料長度：110

In [4]:
x_train = sequence.pad_sequences(x_train, maxlen=110)
x_test = sequence.pad_sequences(x_test, maxlen=110)

# 函數學習機

LSTM=64
Dropout=0.5

In [5]:
model = Sequential()
model.add(Embedding(12000, 128))
model.add(LSTM(64, dropout=0.5, recurrent_dropout=0.5))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy',
             optimizer='adam',
             metrics=['accuracy'])

In [6]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, None, 128)         1536000   
_________________________________________________________________
lstm (LSTM)                  (None, 64)                49408     
_________________________________________________________________
dense (Dense)                (None, 1)                 65        
Total params: 1,585,473
Trainable params: 1,585,473
Non-trainable params: 0
_________________________________________________________________


In [8]:
(64+128+1)*4*64

49408

# 訓練

batch size=30
epochs=5

In [9]:
import os

os.environ['KMP_DUPLICATE_LIB_OK']='True'

In [10]:
model.fit(x_train, y_train, batch_size=30, epochs=5,
         validation_data=(x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 421s 17ms/sample - loss: 0.5102 - accuracy: 0.7508 - val_loss: 0.4217 - val_accuracy: 0.8103
Epoch 2/5
25000/25000 [==============================] - 381s 15ms/sample - loss: 0.3857 - accuracy: 0.8354 - val_loss: 0.3964 - val_accuracy: 0.8311
Epoch 3/5
25000/25000 [==============================] - 394s 16ms/sample - loss: 0.3504 - accuracy: 0.8537 - val_loss: 0.4064 - val_accuracy: 0.8263
Epoch 4/5
25000/25000 [==============================] - 378s 15ms/sample - loss: 0.3123 - accuracy: 0.8737 - val_loss: 0.4363 - val_accuracy: 0.8189
Epoch 5/5
25000/25000 [==============================] - 381s 15ms/sample - loss: 0.2914 - accuracy: 0.8811 - val_loss: 0.4232 - val_accuracy: 0.8266


# 訓練結果

試了以下調整：
Number of words=12,000
資料長度：110
LSTM=64
Dropout=0.5
batch size=30
epochs=5

準確率：0.88